modules/heat_conduction/src/userobjects/GrayLambertSurfaceRadiation.C

```diff
@@ -0,0 +1,302 @@
+                   "fixed_temperature_boundary must be a subset of boundary.");
+      _side_type[_side_id_index.find(p.first)->second] = FIXED_TEMPERATURE;
+    }
+  }
+
+  // get the fixed boundaries of the system if any are provided
+  if (isParamValid("adiabatic_boundary"))
+  {
+    // get the adiabatic boundaries and temperatures
+    std::vector<BoundaryName> boundary_names =
+        getParam<std::vector<BoundaryName>>("adiabatic_boundary");
+    ;
+    for (auto & name : boundary_names)
+      _adiabatic_side_ids.insert(_mesh.getBoundaryID(name));
+
+    // check that adiabatic side ids is a subset of boundary ids
+    // and update _side_type info
+    for (auto & id : _adiabatic_side_ids)
+    {
+      if (_side_id_index.find(id) == _side_id_index.end())
+        paramError("adiabatic_boundary", "adiabatic_boundary must be a subset of boundary.");
+      _side_type[_side_id_index.find(id)->second] = ADIABATIC;
+    }
+
+    // make sure that adiabatic boundaries are not already fixed boundaries
+    for (auto & id : _adiabatic_side_ids)
+      if (_fixed_side_id_index.find(id) != _fixed_side_id_index.end())
+        paramError("adiabatic_boundary", "Isothermal boundary cannot also be adiabatic boundary.");
+  }
+}
+
+void
+GrayLambertSurfaceRadiation::execute()
+{
+  if (_side_id_index.find(_current_boundary_id) == _side_id_index.end())
+    mooseError("Current boundary id: ", _current_boundary_id, " not in _side_id_index.");
+  unsigned int index = _side_id_index.find(_current_boundary_id)->second;
+
+  for (unsigned int qp = 0; qp < _qrule->n_points(); qp++)
+  {
+    _areas[index] += _JxW[qp] * _coord[qp];
+
+    Real temp = 0;
+    if (_side_type[index] == ADIABATIC)
+      continue;
+    else if (_side_type[index] == VARIABLE_TEMPERATURE)
+      temp = _temperature[qp];
+    else if (_side_type[index] == FIXED_TEMPERATURE)
+    {
+      unsigned int iso_index = _fixed_side_id_index.find(_current_boundary_id)->second;
+      temp = _fixed_side_temperature[iso_index]->value(_t, _q_point[qp]);
+    }
+
+    _beta[index] += _JxW[qp] * _coord[qp] * _sigma_stefan_boltzmann * _emissivity[index] *
+                    MathUtils::pow(temp, 4);
+    _side_temperature[index] += _JxW[qp] * _coord[qp] * temp;
+  }
+}
+
+void
+GrayLambertSurfaceRadiation::initialize()
+{
+  for (unsigned int j = 0; j < _n_sides; ++j)
+  {
+    _areas[j] = 0;
+    _beta[j] = 0;
+    _side_temperature[j] = 0;
+  }
+}
+
+void
+GrayLambertSurfaceRadiation::finalize()
+{
+  // need to do some parallel communiction here
+  for (unsigned int j = 0; j < _n_sides; ++j)
+  {
+    _communicator.sum(_areas[j]);
+    _communicator.sum(_beta[j]);
+    _communicator.sum(_side_temperature[j]);
+  }
+
+  // first compute averages from the totals
+  for (unsigned int j = 0; j < _n_sides; ++j)
+  {
+    _beta[j] /= _areas[j];
+    _side_temperature[j] /= _areas[j];
+  }
+
+  // matrix and rhs vector for the view factor calculation
+  DenseMatrix<Real> matrix(_n_sides, _n_sides);
```

Comment:
This matrix assembly can be moved into constructor. If we directly use DenseVector<Real> for `_beta`, we can completely remove this block of codes here.

Reply:
This requires a bit more careful consideration. I may have another user object provide the view factors. This user object must have an opportunity to compute the view factors so we cannot put this into the constructor. 

It could be in one of the setup function or in initialize. I could also check if the matrix needs to be rebuilt or something; think of moving walls e.g.. Maybe `timestepSetup`?